Update iptables perf / KEP-3453 discussion for 1.27 (#39188)
kubernetes/website: content/en/docs/reference/networking/virtual-ips.md

...
```

> ##### Experimental performance improvements {#minimize-iptables-restore}

Reviewer: Are these _still_ experimental?
```suggestion
##### Performance optimizations for `iptables` mode {#minimize-iptables-restore}
```

{{< feature-state for_k8s_version="v1.27" state="beta" >}}

In Kubernetes 1.27, new performance improvements to the iptables proxy
mode are enabled by default, which should make it much less necessary to

> override the default syncing timeouts.

Reviewer 2: should we add context that this is focused on large environments, as in number of clusters or services?

> override the default syncing timeouts.

Reviewer: Try to describe the current state of Kubernetes, post-release:
```markdown
By default, the `kube-proxy` component of Kubernetes {{< skew currentVersion >}}
aggregates its updates to `iptables` rules.
Previous versions of Kubernetes (before v1.27) did not default to this behavior, and
FIXME - what does this mean for a cluster operator?
```

If you were previously overriding `minSyncPeriod`, you should try
removing that override and letting kube-proxy use the default value
(`1s`) or at least a smaller value than you were using before.

If you notice kube-proxy's
`sync_proxy_rules_iptables_restore_failures_total` or
`sync_proxy_rules_iptables_partial_restore_failures_total` metrics
increasing after upgrading to 1.27, that likely indicates you are
encountering bugs in the new code, and you should file a bug report.
(You can disable the new optimizations by disabling the
`MinimizeIPTablesRestore` [feature
gate](/docs/reference/command-line-tools-reference/feature-gates/) for

> kube-proxy with `--feature-gates=MinimizeIPTablesRestore=false,…`.)

Reviewer: We avoid documenting transitions between versions within the reference docs (exception: https://kubernetes.io/docs/reference/using-api/deprecation-guide/ and friends). As docs and features move towards GA, this approach becomes more important.

As a result, this information probably belongs in the post-release blog article _and_ in the v1.27 release notes.
The PR that adds the blog article can also update this page to link there, and we can merge that on release day provided that folks plan ahead and get everything ready for review in plenty of time.

Author: ok, it's already in the release note for https://github.com/kubernetes/kubernetes/pull/115138

Reviewer: The next step is to cut out or reword this section - otherwise, you'll need to commit SIG Network to maintaining this and updating it for the v1.28 release, when this wording will be stale.

...
```

> ##### Performance improvements for `iptables` mode {#minimize-iptables-restore}

Reviewer: ```suggestion
##### Performance optimization for `iptables` mode {#minimize-iptables-restore}
```
(describe the state, not the diff - see https://developers.google.com/style/timeless-documentation for a good explanation on the motivation)

Now it's not an experiment, we should reframe.

{{< feature-state for_k8s_version="v1.27" state="beta" >}}

> In the past (before v1.27), kube-proxy updated all the rules for all
> Services on every sync, which contributed to the slowness in very
> large clusters. For later releases - including Kubernetes {{< skew currentVersion >}} -
> the kube-proxy defaults to an approach that optimizes
> updates better to make smaller iptables updates when only a few
> Services (or their EndpointSlices) have changed. As a result, kube-proxy configuration
> changes
> that were made to improve performance in the past may no longer be
> necessary, and may actually have a negative effect on overall proxy
> performance.

Reviewer: ```suggestion
In Kubernetes {{< skew currentVersion >}} the kube-proxy defaults to a minimal approach
to `iptables-restore` operations, only making updates where Services or EndpointSlices have
actually changed. This is a performance optimization.
The original implementation updated all the rules for all Services on every sync; this
sometimes led to performance issues (update lag) in large clusters.

If you are not running kube-proxy from Kubernetes {{< skew currentVersion >}}, check
the behavior and associated advice for the version that you are actually running.
```

> If you were previously overriding `minSyncPeriod`, you should try
> removing that override and letting kube-proxy use the default value
> (`1s`) or at least a smaller value than you were using before.

Reviewer: ```suggestion
If you were previously overriding `minSyncPeriod`, you should try
removing that override and letting kube-proxy use the default value
(`1s`) or at least a smaller value than you were using before upgrading.
You can select the legacy behavior by disabling the `MinimizeIPTablesRestore`
[feature gate](/docs/reference/command-line-tools-reference/feature-gates/)
(you should not need to).
```

aggregated, but the downside is that each individual change may end up
waiting up to the full `minSyncPeriod` before being processed, meaning
that the iptables rules spend more time being out-of-sync with the
current API server state.

The default value of `1s` should work well in most clusters, but in very
large clusters it may be necessary to set it to a larger value.
(Especially, if kube-proxy's `sync_proxy_rules_duration_seconds` metric
indicates an average time much larger than 1 second, then bumping up

> `minSyncPeriod` may make updates more efficient.)

Reviewer 3: I don't think we need these parenthesis.
